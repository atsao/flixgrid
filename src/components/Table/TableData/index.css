.flixgrid-td-empty {
  text-align: center;
  opacity: 0.3;
  padding: 24px;
}


.flixgrid-td {
  padding: 8px 10px;
}

input[type=text] {
  border: solid 1px #ddd;
  border-radius: 4px;
  margin: 2px auto;
  padding: 4px 8px;
  transition: all 0.2s ease-out;
}

input[type=text]:focus {
  outline: none;
  border-color: #29a9d1;
  box-shadow: inset 0 2px 3px rgba(0, 0, 0, 0.05);
}
