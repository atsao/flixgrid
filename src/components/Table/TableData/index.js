import React, { Component } from 'react';
import PropTypes from 'prop-types';

class TableData extends Component {
  _renderHeader = () => {
    const {
      display,
      dataKey,
      render,
      sort,
      sorting,
      filter,
      filtering,
    } = this.props;
    const content = display || '';

    return (
      <td>
        {render ? render(content) : content}
        <button
          onClick={() => sort(dataKey, 'asc')}
          disabled={sorting.dataKey === dataKey && sorting.direction === 'asc'}
          className="sort-ascending"
        >
          Asc
        </button>
        <button
          onClick={() => sort(dataKey, 'desc')}
          disabled={sorting.dataKey === dataKey && sorting.direction === 'desc'}
          className="sort-descending"
        >
          Desc
        </button>
        <input
          type="text"
          onChange={e => filter(dataKey, e.target.value)}
          className="filter"
        />
      </td>
    );
  };

  _renderEmpty = () => {
    const { data, columns, render } = this.props;
    const content = data || '';

    return (
      <td colSpan={columns.length}>
        {render ? render(content) : content}
      </td>
    );
  };

  render() {
    const { data, dataKey, render, _isHeader, _isEmpty } = this.props;
    const content = data[dataKey] || data || '';

    if (_isHeader) {
      return this._renderHeader();
    }

    if (_isEmpty) {
      return this._renderEmpty();
    }

    return (
      <td>
        {render ? render(content) : content}
      </td>
    );
  }
}

TableData.propTypes = {
  data: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
    PropTypes.string,
  ]),
  columns: PropTypes.array,
  _isHeader: PropTypes.bool,
  _isEmpty: PropTypes.bool,
  display: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  render: PropTypes.func,
  dataKey: PropTypes.string,
  sort: PropTypes.func,
  sorting: PropTypes.object,
};

export default TableData;
