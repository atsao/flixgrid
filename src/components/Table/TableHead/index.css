.flixgrid-thead td {
  background: #eee;
  border-bottom: solid 1px #ddd;
  font-weight: 800;
  padding: 10px;
}

.flixgrid-thead td:first-child,
.flixgrid-thead td:last-child {
  border-top-left-radius: 4px;
}
