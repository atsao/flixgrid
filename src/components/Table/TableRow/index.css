.flixgrid-tr {
  background: #f7f7f7;
}

.flixgrid-tr:nth-child(2n+1) {
  background: #fff;
}

.flixgrid-tr:last-child td:first-child {
  border-bottom-left-radius: 4px;
}

.flixgrid-tr:last-child td:last-child {
  border-bottom-right-radius: 4px;
}
