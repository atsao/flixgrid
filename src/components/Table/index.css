.flixgrid-table {
  border: solid 1px #ddd;
  border-collapse: separate;
  border-radius: 4px;
  border-spacing: 0;
  width: 100%;
}
