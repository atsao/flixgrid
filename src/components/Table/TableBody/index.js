import React from 'react';
import PropTypes from 'prop-types';

import TableRow from '../TableRow';

const TableBody = props => {
  const { data, columns } = props;

  return (
    <tbody>
      {data && data.length
        ? data.map((data, i) => (
            <TableRow key={i} columns={columns} data={data} />
          ))
        : <TableRow columns={columns} data="No data" _isEmpty />}
    </tbody>
  );
};

TableBody.propTypes = {
  data: PropTypes.array,
  columns: PropTypes.array,
};

export default TableBody;
