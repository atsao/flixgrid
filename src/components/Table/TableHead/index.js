import React from 'react';
import PropTypes from 'prop-types';

import TableRow from '../TableRow';

const TableHead = props => {
  const { data, columns, sort, sorting, filter, filtering } = props;

  return (
    <thead>
      <TableRow
        _isHeader
        data={data}
        columns={columns}
        sort={sort}
        sorting={sorting}
        filter={filter}
        filtering={filtering}
      />
    </thead>
  );
};

TableHead.propTypes = {
  data: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
  columns: PropTypes.array,
  sort: PropTypes.func,
  sorting: PropTypes.object,
  filter: PropTypes.func,
  filtering: PropTypes.object,
};

export default TableHead;
