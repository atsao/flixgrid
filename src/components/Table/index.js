import React, { Component } from 'react';
import PropTypes from 'prop-types';

import styles from './index.css';
import TableHead from './TableHead';
import TableBody from './TableBody';

class Table extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sorting: {
        dataKey: '',
        direction: '',
      },
      filtering: {
        dataKey: '',
        query: '',
      },
    };
  }

  _handleSort = (dataKey, direction) => {
    const { sorting } = this.state;
    let newSorting = JSON.parse(JSON.stringify(sorting));
    newSorting.dataKey = dataKey;
    newSorting.direction = direction;

    this.setState({
      sorting: newSorting,
    });
  };

  _handleFilter = (dataKey, query) => {
    const { filtering } = this.state;
    let newSorting = JSON.parse(JSON.stringify(filtering));
    newSorting.dataKey = dataKey;
    newSorting.query = query.toString();

    this.setState({
      filtering: newSorting,
    });
  };

  _sortData = data => {
    const { sorting: { dataKey, direction } } = this.state;

    if (dataKey && direction) {
      let high = 1;
      let low = -1;
      if (direction === 'desc') {
        high = -1;
        low = 1;
      }
      return data.sort((a, b) => a[dataKey] > b[dataKey] ? high : low);
    } else {
      return data;
    }
  };

  _filterData = data => {
    const { filtering: { dataKey, query } } = this.state;
    if (dataKey && query) {
      return data.filter(
        d =>
          d[dataKey].toString().toLowerCase().indexOf(query.toLowerCase()) > -1
      );
    } else {
      return data;
    }
  };

  _renderTableHead = () => {
    const { sorting, filtering } = this.state;

    return (
      <TableHead
        sort={this._handleSort}
        sorting={sorting}
        filter={this._handleFilter}
        filtering={filtering}
        {...this.props}
      />
    );
  };

  render() {
    const { data, columns, showHeaders } = this.props;
    const sortedData = this._sortData(data);
    const filteredData = this._filterData(sortedData);

    return (
      <table className={styles['flixgrid-table']}>
        {showHeaders && this._renderTableHead()}
        <TableBody
          data={filteredData}
          columns={columns}
          showHeaders={showHeaders}
        />
      </table>
    );
  }
}

Table.propTypes = {
  data: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
  columns: PropTypes.array,
  showHeaders: PropTypes.bool,
};

export default Table;
