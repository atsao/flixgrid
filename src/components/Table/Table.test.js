import React from 'react';
import renderer from 'react-test-renderer';

import Table from '../Table';
import TableHead from '../Table/TableHead';
import TableBody from '../Table/TableBody';
import TableRow from '../Table/TableRow';
import TableData from '../Table/TableData';

const props = {
  data: [],
  columns: [],
};

it('renders a Table component', () => {
  const tree = renderer.create(<Table {...props} />);
  expect(tree).toMatchSnapshot();
});

it('renders a TableHead component', () => {
  const tree = renderer.create(<TableHead {...props} />);
  expect(tree).toMatchSnapshot();
});

it('renders a TableBody component', () => {
  const tree = renderer.create(<TableBody {...props} />);
  expect(tree).toMatchSnapshot();
});

it('renders a TableRow component', () => {
  const tree = renderer.create(<TableRow {...props} />);
  expect(tree).toMatchSnapshot();
});

it('renders a TableData component', () => {
  const tree = renderer.create(<TableData {...props} />);
  expect(tree).toMatchSnapshot();
});
