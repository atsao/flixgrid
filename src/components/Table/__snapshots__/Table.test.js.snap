// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Table component 1`] = `
<table>
  <tbody>
    <tr>
      <td
        colSpan={0}
      >
        No data
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`renders a TableBody component 1`] = `
<tbody>
  <tr>
    <td
      colSpan={0}
    >
      No data
    </td>
  </tr>
</tbody>
`;

exports[`renders a TableData component 1`] = `<td />`;

exports[`renders a TableHead component 1`] = `
<thead>
  <tr />
</thead>
`;

exports[`renders a TableRow component 1`] = `<tr />`;
